fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/RecordQueryIndexPlan.java

```diff
@@ -84,37 +100,103 @@ public class RecordQueryIndexPlan implements RecordQueryPlanWithNoChildren, Reco
+            }
+            if (scanParameters.getScanType() != IndexScanType.BY_VALUE) {
+                error("Index Prefetch can only be used with VALUE index scan. Falling back to regular scan.");
+                this.useIndexPrefetch = RecordQueryPlannerConfiguration.IndexPrefetchUse.NONE;
+            }
+            if (!FDBDatabaseFactory.instance().getDatabase().getAPIVersion().isAtLeast(APIVersion.API_VERSION_7_1)) {
+                error("Index Prefetch can only be used with API_VERSION of at least 7.1. Falling back to regular scan.");
+                this.useIndexPrefetch = RecordQueryPlannerConfiguration.IndexPrefetchUse.NONE;
+            }
+        }
+    }
+
+    @Nonnull
+    @Override
+    public <M extends Message> RecordCursor<QueryResult> executePlan(@Nonnull final FDBRecordStoreBase<M> store, @Nonnull final EvaluationContext context,
+                                                                     @Nullable final byte[] continuation, @Nonnull final ExecuteProperties executeProperties) {
+        switch (useIndexPrefetch) {
+            case NONE:
+                // Use the default implementation for index scan
+                return RecordQueryPlanWithIndex.super.executePlan(store, context, continuation, executeProperties);
+            case USE_INDEX_PREFETCH:
+                // Use index prefetch without fallback
+                return executeUsingIndexPrefetch(store, context, continuation, executeProperties);
+            case USE_INDEX_PREFETCH_WITH_FALLBACK:
+                // Use Index prefetch and fall back to regular index scan.
+                // Using the fallback mechanism here separates the execution part from the planning part
+                // (No need to plan again) and from failures at other parts of the execution.
+                // In practice, the same continuation should not be used for both kinds of cursors. In practice, we
+                // rely on the fact that the fallback mode should not be used with a non-empty continuation:
+                // If the previous call succeeded, we should have PREFETCH mode and if it failed, it should be NONE
+                try {
+                    // The fallback cursor will handle failures that happen after the executeUsingIndexPrefetch call
+                    return new FallbackCursor<>(
+                            executeUsingIndexPrefetch(store, context, continuation, executeProperties),
+                            () -> RecordQueryPlanWithIndex.super.executePlan(store, context, continuation, executeProperties));
+                } catch (Exception ex) {
+                    KeyValueLogMessage message = KeyValueLogMessage.build("Index Prefetch plan failed, falling back to Index scan",
+                            LogMessageKeys.PLAN_HASH, planHash(PlanHashKind.STRUCTURAL_WITHOUT_LITERALS));
+                    LOGGER.error(message.toString(), ex);
```

Comment:
`ERROR` seems pretty strict for this log line. Given that query execution succeeds, it seems like `WARN` is the highest this should be, and it maybe even should be `INFO` if we expect this kind of thing to happen in normal operation (e.g., if the user calls this with a snapshot transaction).

Reply:
Changed to `WARN` since I thought it would be good to be notified when we have a fallback case, though once the feature is rolled out we may want to convert to `INFO` to reduce chattiness.

Reply:
Is this missing a push? I'm still seeing `error` in GitHub

Reply:
Escaped error that is now fixed (to a warning)